app/assets/javascripts/discourse/app/static/prosemirror/core/plugin.js

```diff
@@ -1,15 +1,34 @@
 import { Plugin } from "prosemirror-state";
 
+/*
+  There are 3 ways to define a node view:
+
+  Setting a node view class directly (e.g. code-block)
+  `nodeViews: { nodeName: NodeViewClass }`
+
+  Setting a node view class as returned by a function, when plugin params are needed (e.g. footnote)
+  `nodeViews: { nodeName: (pluginParams) => NodeViewClass }`
+
+  Setting a node view instance as returned by a function, when plugin params are needed (e.g. image)
+  `nodeViews: { nodeName: (pluginParams) => (...args) => nodeViewInstance }`
+*/
 export function extractNodeViews(extensions, pluginParams) {
   /** @type {Record<string, import('prosemirror-view').NodeViewConstructor>} */
   const allNodeViews = {};
   for (const { nodeViews } of extensions) {
     if (nodeViews) {
-      for (let [name, NodeViewClass] of Object.entries(nodeViews)) {
-        if (!NodeViewClass.toString().startsWith("class")) {
-          NodeViewClass = NodeViewClass(pluginParams);
+      for (let [name, nodeView] of Object.entries(nodeViews)) {
+        // node view can be a function, to which we pass pluginParams
+        if (!nodeView.toString().startsWith("class")) {
```

Comment:
Insane that this is the best way to check if something is a class and not a function in JS 😆

Reply:
Yeah it's so ugly, I'd be happy to change it... I tried inferring from .prototype.constructor but reverted back to this way for now